grails-app/services/org/pih/warehouse/data/LoadDataService.groovy

```diff
@@ -0,0 +1,291 @@
+
+            csvItems.set(i, newItem);
+        }
+
+        ImportDataCommand command = new ImportDataCommand();
+
+        command.setData(csvItems);
+
+        productSupplierDataService.validate(command)
+        productSupplierDataService.process(command)
+
+        csvReader.close();
+    }
+
+
+    def importInventory(URL csvURL) {
+        CSVMapReader csvReader = new CSVMapReader(csvURL.newInputStream().newReader());
+
+        csvReader.eachLine { Map attr ->
+            Product product = Product.findByProductCode(attr["Product code"]);
+
+            InventoryItem inventoryItem = new InventoryItem(
+                    product: product,
+                    lotNumber: attr["Lot number"],
+                    expirationDate: attr["Expiration date"],
+                    comments: attr["Comment"]
+            );
+
+            inventoryItem.save();
+        }
+
+        csvReader.close();
+    }
+
+    def importInventoryLevels(URL csvURL) {
+        CSVMapReader csvReader = new CSVMapReader(csvURL.newInputStream().newReader());
+
+        csvReader.eachLine { Map<String, String> attr ->
+            Product product = Product.findByProductCode(attr["Product code"]);
+
+            Location preferredBinLocation = Location.findByName(attr["Preferred bin location"])
+
+            InventoryLevel inventoryLevel = new InventoryLevel(
+                    product: product,
+                    status: attr["status"],
+                    preferredBinLocation: preferredBinLocation,
+                    minQuantity: attr["Min quantity"],
+                    maxQuantity: attr["Max quantity"],
+                    reorderQuantity: attr["Reorder quantity"],
+                    expectedLeadTimeDays: attr["Expected Lead Time Days"],
+                    replenishmentPeriodDays: attr["Replenishment Period Days"]
+            )
+
+            inventoryLevel.save(failOnError: true);
+        }
+
+        csvReader.close();
+    }
+
+    def importUsers(URL csvURL) {
+        CSVMapReader csvReader = new CSVMapReader(csvURL.newInputStream().newReader());
+
+        csvReader.eachLine { Map<String, String> attr ->
+            StringTokenizer tokenizer = new StringTokenizer(attr.roleType, ",");
+            Set<RoleType> roleTypes = new HashSet<RoleType>();
+
+            while (tokenizer.hasMoreElements()) {
+                roleTypes.add(RoleType.valueOf(tokenizer.nextElement() as String));
+            }
+
+            Set<Role> userRoles = new HashSet<Role>();
+
+            for (RoleType roleType in roleTypes) {
+                Role role = Role.findByRoleType(roleType)
+
+                if (!role) {
+                    role = new Role(
+                            roleType: roleType,
+                            name: roleType.toString(),
+                    )
+                }
+                userRoles.add(role)
+            }
+
+            if (User.findByUsername(attr.username)) {
```

Comment:
This should be already handled through the unique constraint on the User domain.

Reply:
You just need to check if the save was successful and throw an exception if not. 
```
// The hasErrors checks whether there were errors when data binding
if (user.hasErrors() || !user.save()) { 
    throw new ValidationException("invalid user", user.errors)
}
```

Or you can explcitly validate the domain before saving. 
```
user.validate() 
if (user.hasErrors() || !user.save()) { 
    throw new ValidationException("invalid user", user.errors)
}
```
I prefer the first approach.

Reply:
Added failOnError when saving user